Add state dependent icons to moon sensor

Mariusthvdb opened the pull request:
Material design icons have icons for all sensor.states, let's use these natively in the component.
Based on the Season icons, tried to change accordingly.

## Breaking Change:

<!-- What is breaking and why we have to break it. Remove this section only if it was NOT a breaking change. -->

## Description:


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ ] There is no commented out code in this PR.
  - [ ] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

Mariusthvdb committed:
have sensor.moon use state dependent icons

Material design icons have icons for all sensor.states, let's use these natively in the component.
Based on the Season icons, tried to change accordingly.

fabaff commented on homeassistant/components/moon/sensor.py:
```diff
@@ -13,7 +13,16 @@
+    STATE_WANING_GIBBOUS: "moon-waning-gibbous",
+    STATE_LAST_QUARTER: "mdi:moon-last-quarter",
+    STATE_WANING_CRESCENT: "mdi:moon-waning-crescent"
+}
```
The state constants are not defined.

Mariusthvdb replied:
since this is my first pr, I based myself on season.sensor, and missed that , sorry.

would this be what is needed?

```
STATE_NEW_MOON: "new_moon",
STATE_WAXING_CRESCENT: "waxing_crescent",
STATE_FIRST_QUARTER: "first_quarter",
STATE_WAXING_GIBBOUS: "waxing_gibbous",
STATE_FULL_MOON: "full_moon",
STATE_WANING_GIBBOUS: "waning_gibbous",
STATE_LAST_QUARTER: "last_quarter",
STATE_WANING_CRESCENT
```

hope you will help me get to understand all this...

Mariusthvdb replied:
added state constants

luca-angemi replied:
I think this won't work as if I'm not mistaken self._state contains a number (i.e. the number of days after a new moon mod 28), not the state names.

Mariusthvdb replied:
yes, thats what I thought hence my own expressed doubt about the correct call. need to get to the state(self) but don't know how to formulate the final call for that.
please help me?

luca-angemi replied:
I think you could return MOON_ICONS[state] but I'm no expert at all, maybe @fabaff could suggest!

fabaff replied:
`self.state` returns a string which could be used for the lookup.

Now looking at the code, especially at the naming of the states, I would say that my initial idea was to delegate the selection of the icon to the frontend.

Mariusthvdb replied:
thanks for getting back.
to finalize this PR then, what would I need to use for the final line:

```
return MOON_ICONS.get(self.state, "mdi:brightness-3")
```
?
not really sure what you mean by delegating this to the frontend though, seems to me making this native in the backend is preferable to any kind of templating I the frontend (which is what I do know...)
```    sensor.moon:
      templates:
        icon: >
          if (state === 'full_moon') return 'mdi:moon-full';
          if (state === 'new_moon') return 'mdi:moon-new';
          return 'mdi:moon-' + state.replace('_','-');
```

Mariusthvdb committed:
Update sensor.py

added state constants

luca-angemi commented on homeassistant/components/moon/sensor.py:
```diff
@@ -13,7 +13,25 @@
+    STATE_FIRST_QUARTER: "mdi:moon-first-quarter",
+    STATE_WAXING_GIBBOUS: "mdi:moon-waxing-gibbous",
+    STATE_FULL_MOON: "mdi:moon-full",
+    STATE_WANING_GIBBOUS: "moon-waning-gibbous",
```
You forgot an "mdi:" here

Mariusthvdb replied:
added, thanks, good catch...

Mariusthvdb committed:
Update sensor.py

fixed missing mdi:

fabaff commented on homeassistant/components/moon/sensor.py:
```diff
@@ -13,7 +13,25 @@
+STATE_FULL_MOON = "full_moon"
+STATE_WANING_GIBBOUS = "waning_gibbous"
+STATE_LAST_QUARTER = "last_quarter"
+STATE_WANING_CRESCENT = "waning_crescent"
```
Please replace the strings for `self.state` with the constants.

Mariusthvdb replied:
sorry, but I dont understand what this means. Please give me an example, and I will follow suit with the rest. thanks.

luca-angemi replied:
I think he means in the state function where it returns for example

```python 
return "new_moon"
```

you should change this for 
```python 
return STATE_NEW_MOON
```

and so on for all the other states.

Mariusthvdb replied:
I see, thanks, so like this:
```
    @property
    def state(self):
        """Return the state of the device."""
        if self._state == 0:
            return "STATE_NEW_MOON"
        if self._state < 7:
            return "STATE_WAXING_CRESCENT"
        if self._state == 7:
            return "STATE_FIRST_QUARTER"
        if self._state < 14:
            return "STATE_WAXING_GIBBOUS"
        if self._state == 14:
            return "STATE_FULL_MOON"
        if self._state < 21:
            return "STATE_WANING_GIBBOUS"
        if self._state == 21:
            return "STATE_LAST_QUARTER"
        return "STATE_WANING_CRESCENT"
```

?

Mariusthvdb replied:
done

luca-angemi replied:
without quotes

Mariusthvdb replied:
removed....

fabaff reviewed (commented):
Run `black` locally before committing to avoid [style issues](https://developers.home-assistant.io/docs/en/development_guidelines.html).

Mariusthvdb committed:
order of constants. moon icons and use self.state

change order of constants to alphabetical, changed order in Moon_icons to alphabetical, used self.state for icon lookup

Mariusthvdb committed:
replace the strings

 replace the strings for self.state with the constants

Mariusthvdb committed:
removed quotes

Mariusthvdb commented:
so other than running 'black' this is ok now?

Mariusthvdb commented:
need some assistance in how to check this in 'Black'. The link provided to the development_guidelines doesn't really help me tbh... Sorry for my noob-ness

springstan commented:
@Mariusthvdb you need to format it with [black](https://github.com/psf/black) e.g. `black {source_file_or_directory}`

Mariusthvdb commented:
ok so I installed black locally and made a local copy of the pr'd  file.

it errors out on an unaltered line, that has been there before my effort:

```
$ black /Users/theboss/Desktop/sensor.py 
error: cannot format /Users/theboss/Desktop/sensor.py: Cannot parse: 50:0:      def state(self):
Oh no! 💥 💔 💥
1 file failed to reformat.
```

now what. Really sorry, but I am afraid without reasonable instructions how to setup a local PR-test configuration this is all a bit too much for me...

springstan commented:
@Mariusthvdb have you [set up a dev environment](https://developers.home-assistant.io/docs/en/development_environment.html) or are you working without one?
Because inside of a dev environment you can simply run `black --fast homeassistant` to format your code and in addition `pre-commit` checks that your code is formatted with `black` whenever you commit. 
See [this](https://developers.home-assistant.io/blog/2019/07/31/black.html) dev blog post.

Mariusthvdb commented:
> @Mariusthvdb have you [set up a dev environment](https://developers.home-assistant.io/docs/en/development_environment.html) or are you working without one?
> Because inside of a dev environment you can simply run `black --fast homeassistant` to format your code and in addition `pre-commit` checks that your code is formatted with `black` whenever you commit.
> See [this](https://developers.home-assistant.io/blog/2019/07/31/black.html) dev blog post.

thanks, I am working without one. Read the page before, and, using a Mac, installed home-brew, but after that, well, gave up really. 
Hope someone else will takeover this PR, and manages to finish it.

springstan commented on homeassistant/components/moon/sensor.py:
```diff
@@ -13,7 +13,26 @@
+    STATE_FULL_MOON: "mdi:moon-full",
+    STATE_LAST_QUARTER: "mdi:moon-last-quarter",
+    STATE_NEW_MOON: "mdi:moon-new",
+    STATE_WANING_CRESCENT: "mdi:moon-waning-crescent",   
```
I looked at the error and it seems like you have spaces at the end of the line which need to be removed.
See https://dev.azure.com/home-assistant/Home%20Assistant/_build/results?buildId=16325&view=logs&j=4e214fed-f310-53fa-0018-b38e7beb405b&t=4e7ef364-9046-5187-cd06-98a96bf53c68&l=21

springstan commented on homeassistant/components/moon/sensor.py:
```diff
@@ -13,7 +13,26 @@
+    STATE_WANING_GIBBOUS: "mdi:moon-waning-gibbous",
+    STATE_WAXING_CRESCENT: "mdi:moon-waxing-crescent",
+    STATE_WAXING_GIBBOUS: "mdi:moon-waxing-gibbous"
+    
```
And remove this blank line, error on Azure DevOps: https://dev.azure.com/home-assistant/Home%20Assistant/_build/results?buildId=16325&view=logs&j=4e214fed-f310-53fa-0018-b38e7beb405b&t=4e7ef364-9046-5187-cd06-98a96bf53c68&l=22

Mariusthvdb replied:
thanks!

springstan replied:
No problem! Thanks for contributing ;-)

Mariusthvdb replied:
still 2 tests fail...and it still shows fabaff requested changed. Which I think I made. So don't really understand what's left to do, other than rerun the test?

springstan replied:
Well not exactly only one test fails which results in the entire build failing. Yes you made the requested changes, now you just have to wait that they are approved. I cannot figure out what the error of the CheckFormat really means, so I would also suggest to rerun it.

Mariusthvdb replied:
Ok thanks, I'll wait, and in the mean time was able to run Black locally, which is now saved and tested. Hope this clinches it then! thanks for your support!

springstan reviewed (changes requested):
This should fix your formatting errors :)

Mariusthvdb committed:
removed spaces, empty line

Mariusthvdb committed:
local Black

Mariusthvdb commented:
not sure what the only failing check requires me to do now, I'll await further instructions.
file should be Black now, and all other requested changes are in place.

Mariusthvdb commented:
anyone can help with the codecov/project issue? final change to make things Black only added 1 comma, and suddenly this error appears...

fabaff committed:
Sort imports and move import to file header

Mariusthvdb commented:
Anyone on why this is still unsuccessful ?

frenck commented:
@Mariusthvdb The test coverage of this integration dropped, because you've added lines of code.

<https://github.com/home-assistant/home-assistant/blob/dev/tests/components/moon/test_sensor.py>

Please add some additional unit tests to keep this integration.

Mariusthvdb commented:
o dear...
this only happened after I replaced the original code with the blackified code. And believe it or not, the only extra character was the final comma......

having said that, I will certainly have a look if I can add a test, I take it for the icon?

frenck commented:
Hmm.. looking at it, you've actually increased the coverage with 2.66%. So it's fine I guess.

The main issue (looking at codecov), is not the icons, but the other moon phases not being tested.

Considering that, this is all out of scope for this PR. So let's go ahead and merge this in.

If you are up for a little challenge and want to learn more about testing, I can definitely recommend this integration, since it is fairly simple and a good starter, with already some tests in place that simply need to be extended.

Mariusthvdb commented:
thanks Frenck, cool.
And will do. Hope to learn. Need to.....
